File: examples/governance/compliance_domain/lib/compliance_domain/training_record/commands/renew_training.rb
module ComplianceDomain
  class TrainingRecord
    module Commands
      class RenewTraining
        include Hecks::Command
        emits "RenewedTraining"

        attr_reader :training_record_id
        attr_reader :certification_id
        attr_reader :expires_at

        def initialize(
          training_record_id: nil,
          certification_id: nil,
          expires_at: nil
        )
          @training_record_id = training_record_id
          @certification_id = certification_id
          @expires_at = expires_at
        end

        def call
          existing = repository.find(training_record_id)
          if existing
            unless existing.status == "completed"
              raise Hecks::Error, "Cannot RenewTraining: status must be 'completed', got '#{existing.status}'"
            end
            TrainingRecord.new(
              id: existing.id,
              stakeholder_id: existing.stakeholder_id,
              policy_id: existing.policy_id,
              expires_at: expires_at,
              certification_id: certification_id,
              completed_at: Time.now.to_s,
              status: "completed"
            )
          else
            raise Hecks::Error, "TrainingRecord not found: #{training_record_id}"
          end
        end
      end
    end
  end
end

File: examples/governance_static_go/domain/renew_training.go
package domain

import (
	"time"
	"fmt"
)

type RenewTraining struct {
	TrainingRecordId string `json:"training_record_id"`
	Certification string `json:"certification"`
	ExpiresAt time.Time `json:"expires_at"`
}

func (c RenewTraining) CommandName() string { return "RenewTraining" }

func (c RenewTraining) Execute(repo TrainingRecordRepository) (*TrainingRecord, *RenewedTraining, error) {
	existing, err := repo.Find(c.TrainingRecordId)
	if err != nil {
		return nil, nil, err
	}
	if existing == nil {
		return nil, nil, fmt.Errorf("TrainingRecord not found: %s", c.TrainingRecordId)
	}
	existing.Certification = c.Certification
	existing.ExpiresAt = c.ExpiresAt
	if existing.Status != "completed" {
		return nil, nil, fmt.Errorf("cannot RenewTraining: TrainingRecord is in %s state", existing.Status)
	}
	existing.Status = "completed"
	existing.UpdatedAt = time.Now()
	if err := existing.Validate(); err != nil {
		return nil, nil, err
	}
	if err := repo.Save(existing); err != nil {
		return nil, nil, err
	}
	event := RenewedTraining{
		AggregateID: existing.ID,
		TrainingRecordId: c.TrainingRecordId,
		Certification: c.Certification,
		ExpiresAt: c.ExpiresAt,
		OccurredAt: time.Now(),
	}
	return existing, &event, nil
}
